### DISCUSSION TITLE
Can you tell me how to ensure high performance of writing to the stream?

### DISCUSSION THREAD

**Discussion by mshakurov** [SixLabors/ImageSharp#2572] [Q&A] [answered]
We noticed that saving the images we created (png format) to MemoryStream for some encoder settings is slow for our tasks.
The simplest tests show <del>amazing things</del>.

For example, download the map tile obtained from OSM at https://tile.openstreetmap.org/6/11/17.png .
This will be a 256*256 png image, file size 13,829.

Let's upload this file in a simple way: `var img = Image.Load( fileName );`

Then, let's look at some img instance parameters: _Metadata.HorizontalResolution: 96 x 96, Metadata.ResolutionUnits: PixelsPerInch, PixelType.BitsPerPixel BitsPerPixel: 32._

And now let's try to save this instance to `MemoryStream` using the simple `SaveAsPng` method:
```
var encoder = new PngEncoder { BitDepth = PngBitDepth.Bit4, ColorType = PngColorType.Palette };
using Stream ms = new MemoryStream( /*capacity*/ );
img.SaveAsPng( ms, encoder );
```
This saving works very slowly.
Regardless of `PngBitDepth`, if `ColorType` is set to `PngColorType.Palette`, then saving is very slow.
If you do not use `Palette`, but, for example, `RgbWithAlpha`, then the recording will be much faster, but the size of the stream also grows significantly.

**The main question:**
What advice would you give to get a very **high speed saving**  of the image from the example: a) to MemoryStream **in png format**, b) with **minimal size increase**?

**Comment by JimBobSquarePants** [comment]
What do you mean by very slow?
ImageSharp is considered fast in comparison to other libraries. In your example you're asking the library to reduce the color palette count to enable a few colors. This is an intensive operation.

**Reply by mshakurov** [reply]
I'm sorry that I couldn't describe the task and the problem in more detail.

You note that we are lowering the quality. But this is not the goal, I just gave this example without explanation (I apologize): the loading an image from the Internet and saving it "as is" so that it is not too big compared to the original one. For example, if you download an image (from the example) in the browser, as I wrote, it offers to be saved as png, and it will turn out to be _13,829_ bytes in size. If the stream of the same image received from the Internet is loaded like this: `Image.Load( responseStream )`, and then saved as a _png_ by method `Image.SaveAsPng(...)`, then it is not clear to us how to achieve a saved image size similar to 14KB and an acceptable save speed.

Therefore, we have a question, how do we use `ImageSharp` for the following purposes:
1) download an image from the Internet and store it as something representing an image, with minimal loss in size/quality and minimal loss in saving speed;
2) upon request, very quickly return unchanged images in the form of a `Stream`. You can suggest storing a `Stream` previously prepared (or a `byte array`), and this is understandable, it is acceptable, but then see the next paragraph;
3) apply shapes/texts to images from memory on requests (this works quickly in `ImageSharp`), and very quickly return the modified images as a `Stream` (this works slowly when using some encoders because we don't know how to get the images "as is" from the downloaded ones). You can also suggest storing modified copies, as is done in your `ImageSharp.Web` cache. But we have a very wide variety of changes, reuse of modified images is not required at all, which means there is no point in using a cache for modified copies.

Perhaps all these questions will turn to such:
- how to convert a downloaded image, without loss/increase in quality, quickly into an `Image` instance;
- how to quickly turn an `Image` into a `Stream` after changing it;

PS: Previously we worked with _gdi+_, we have our own library/service. This library/service, upon request, applies various graphics (texts, images, shapes) to copies of images stored in memory, then puts a copy of the image into the stream and, as a result, returns the stream to the consumer of our service/library.
The performance loss when writing the processed image to the stream compared to the raw one was very small in _gdi+_, the image quality did not decrease much, and the size of the images did not increase so significantly. In addition, saving to Stream worked very quickly. It is clear that this is due to a complete difference in the methods for storing images in memory and saving them to the stream in _gdi+_ and `ImageSharp`. And it is clear that we are incorrectly using all the richness and capabilities of `ImageSharp`.

**Reply by tocsoft** [reply]
what are you using to measure it 'being slow'? what's your runtime environment (.net version, image sharp version etc).

Its important to know that Image Sharp make heavy use of hardware intrinsic/simd for its speed, falling back to much much slower paths when not available, which (I believe) is the case in typical debug builds as the .net runtime does not enable them in that case.... so you will need to make sure you only measure image sharp speed issues when the application code is built/running in release mode with optimizations enabled (on by default in release builds, and disabled in debug builds).

**Reply by mshakurov** [reply]
Of course, we tested the performance in a variety of ways on different platforms.
Of course, we compared the recording performance in MemoryStream, FileStream for different encoder settings. 
On all physical platforms, in all tests, the difference in performance is proportionally the same for different encoder settings.
Some settings give excellent performance. And some settings do not suit us in terms of performance because we use `ImageSharp` incorrectly. 
But the author of the first answer gives me a reasonable answer - the operation of reducing / increasing quality is a costly operation. We agree with this 100%.
But we need to upload some abstract image to the Image instance so that later it is saved in **approximately the same quality** as it was loaded, with **approximately the same size**, and so that it is **saved quickly**.

**Reply by JimBobSquarePants** [reply]
Are the input images pngs? 
Are they 24 or 32 bits per pixel? 
Can you share some example code?

**Reply by mshakurov** [reply]
- lets consider - they all png
- I'm not a guru in graphics, but your library shows the following:
PixelsPerInch - 96 x 96
BitsPerPixel: 32
- I will prepare a very simple test, and I will post it here

**Reply by JimBobSquarePants** [reply]
Thanks. A test image would be great also.

**Reply by mshakurov** [reply]
[TestImageSharpNet6.zip](https://github.com/SixLabors/ImageSharp/files/13171536/TestImageSharpNet6.zip)

A simple console application... Test image is included, but the application can download it from OSM tile web service.

**Reply by JimBobSquarePants** [reply, marked as answer]
Thanks. 

Looks like the images are palette based which will always require quantization. If you test using the new v3.1 alpha (_see our readme for the feed details_) we have optimized the pipeline to reuse the same palette if available. That should be much faster. 

For benchmarking you really shouldn't use anything other than [BenchmarkDotnet](https://benchmarkdotnet.org/). Anything else will be wildly inaccurate.
